packages/library-base/src/commonMain/kotlin/io/realm/DynamicRealmObject.kt

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2022 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm
+
+import kotlin.reflect.KClass
+
+/**
+ * A **dynamic realm object** gives access the data of the realm objects through a generic string
+ * based API instead of the conventional [Realm] API that only allows access through the properties
+ * of the corresponding schema classes supplied in the configuration.
+ */
+interface DynamicRealmObject : RealmObject {
+    /**
+     * The name of the class in the realm object model.
+     */
+    val type: String
+
+    /**
+     * Returns the value of a specific value property.
+     *
+     * *NOTE:* This shouldn't be used for nullable properties. Use [getNullable] for those.
+     *
+     * @param propertyName the name of the property to retrieve the value for.
+     * @param clazz the Kotlin class of the value.
+     * @param T the type of the value.
+     * @return the property value.
+     * @throws IllegalArgummentException if the class doesn't contain a field with the specific
+     * name, or if trying to retrieve collection properties.
+     * @throws ClassCastException if the field doesn't contain a field of the defined return type.
+     */
+    fun <T : Any> get(propertyName: String, clazz: KClass<T>): T
```

Comment:
The documentation doesn't mention how to work with object references, which is probably a bit confusing. So as a minimum we should describe how to get object references.

But that got me thinking....would it make sense for us to differentiate between primitive values and object references....that would also make it easier for us to eventually introduce mutable variants of things...e.g. something like:

```
getValue(propertyName, clazz)
getNullableValue(propertyName, clazz)
getObject(propertyName): DynamicRealmObject? (can be overridden to MutableRealmObject?)
getValueList(propertyname, clazz): DynamicRealmList (can be overridden to MutableDynamicRealmList)
getNullableValueList(propertyName, clazz): DynamicRealmLisna
getObjectList(propertyName): DynamicRealmList<DynamicRealmObject> (Can we override to `MutableDynamicRealmList<MutableDynamicRealmObject>)
```

I'm starting to lean towards that making a cleaner separation where it is easy to see when you move around in the object graph, but we isolate annoying corner cases for objects / lists / maps / sets to their own methods.

Maybe we should meet to discuss?

Reply:
I have updated it according to the above proposal and discussions.